cryptoservice/crypto_service.go

```diff
@@ -82,10 +82,15 @@ func (cs *CryptoService) GetPrivateKey(keyID string) (k data.PrivateKey, role st
 	for _, ks := range cs.keyStores {
 		for _, keyPath := range keyPaths {
 			k, role, err = ks.GetKey(keyPath)
-			if err != nil {
+			if err == nil {
+				return
+			}
+			switch err.(type) {
+			case trustmanager.ErrPasswordInvalid, trustmanager.ErrAttemptsExceeded:
```

Comment:
To the more general point:

The UI for listing all available key stores, letting the user chose between them, telling the user what store will be used in what order, etc., with its formatting, sorting, possibly interactivity, really does not belong in the middle-of-the-call-stack `CryptoService.GetPrivateKey`; it can not even be done in there at all because `CryptoService` does not have any UI channel available (not even `passphrase.Retriever`, which would not be sufficient anyway¹).

If selection between key stores redundantly storing the same key, or even telling the user what the options and impact are, should have any UI, that UI would need to exist somewhere near the top level, and this would impact the whole call stack by having to use (`KeyService`, key ID) as an identifier instead of a key ID stand-alone and having `GetPrivateKey` automatically figure out which `KeyService` the user meant.

(Even if `GetPrivateKey` could always guess right and we only wanted to add `GetWhichServiceWillBeUsedByGetPrivateKey` to allow the top level to have a nicer UI, we should still pass an explicit `KeyService` back to `GetPrivateKey` because the guessing in `GetPrivateKey` depends on changing factors like the user being able to enter the right passphrase or possible storage/ YubiKey hardware failures, so the answer could change between `GetWhichService…` and `GetPrivateKey`.)

¹ (FWIW I would advise against PAM-like (give the user a message, prompt the user for an answer) “UI abstraction” callback: it is inherently untyped/losing information and it makes it impossible to build nice/intelligent UIs by enforcing a sequence-of-those-interactions paradigm.)